rafthttp/util.go

```diff
@@ -45,3 +49,73 @@ func readEntryFrom(r io.Reader, ent *raftpb.Entry) error {
 	}
 	return ent.Unmarshal(buf)
 }
+
+// compareMajorMinorVersion returns an integer comparing two versions based on
+// their major and minor version. The result will be 0 if a==b, -1 if a < b,
+// and 1 if a > b.
+func compareMajorMinorVersion(a, b *semver.Version) int {
+	na := &semver.Version{Major: a.Major, Minor: a.Minor}
+	nb := &semver.Version{Major: b.Major, Minor: b.Minor}
+	switch {
+	case na.LessThan(*nb):
+		return -1
+	case nb.LessThan(*na):
+		return 1
+	default:
+		return 0
+	}
+}
+
+// serverVersion returns the server version from the given header.
+func serverVersion(h http.Header) *semver.Version {
+	verStr := h.Get("X-Server-Version")
+	// backward compatibility with etcd 2.0
+	if verStr == "" {
+		verStr = "2.0.0"
+	}
+	return semver.Must(semver.NewVersion(verStr))
+}
+
+// serverVersion returns the min cluster version from the given header.
+func minClusterVersion(h http.Header) *semver.Version {
+	verStr := h.Get("X-Min-Cluster-Version")
+	// backward compatibility with etcd 2.0
+	if verStr == "" {
+		verStr = "2.0.0"
+	}
+	return semver.Must(semver.NewVersion(verStr))
+}
+
+// checkPeerVersionCompatibility checks whether the given version is compatible
+// with the local version.
+func checkPeerVersionCompatibility(server, minCluster *semver.Version) error {
+	localServer := semver.Must(semver.NewVersion(version.Version))
+	localMinCluster := semver.Must(semver.NewVersion(version.MinClusterVersion))
+	if compareMajorMinorVersion(server, localMinCluster) == -1 {
+		return fmt.Errorf("remote version is too low (remote=%s local=%s)", server, localServer)
+	}
+	if compareMajorMinorVersion(minCluster, localServer) == 1 {
+		return fmt.Errorf("local version is too low (remote=%s local=%s)", server, localServer)
+	}
+	return nil
+}
+
+// isValidClusterVersionString checks whether the given cluster version string
+// is a valid one in the view of local etcd. It is valid when it is an
+// undecided one, or it is +/-1 of the local version.
+func isValidClusterVersionString(verStr string) bool {
```

Comment:
`checkClusterVersionCompatiblity`?
Can we try to make it consistent with the Version one?